File: timeperiod/timeperiod_test.go
package timeperiod

import (
	reflect "reflect"
	"testing"
)

func Test_Parse(t *testing.T) {
	tests := []struct {
		expr    string
		want    TimePeriod
		wantErr bool
	}{
		{
			expr: "this month",
			want: TimePeriod{
				Period:   "month",
				Offset:   0,
				Value:    1,
				Complete: true,
			},
		},
		{
			expr: "this    year  ",
			want: TimePeriod{
				Period:   "year",
				Offset:   0,
				Value:    1,
				Complete: true,
			},
		},
		{
			expr: "next complete day",
			want: TimePeriod{
				Period:   "day",
				Offset:   1,
				Value:    1,
				Complete: true,
			},
		},
		{
			expr: "  next day",
			want: TimePeriod{
				Period:   "day",
				Offset:   0,
				Value:    1,
				Complete: false,
			},
		},
		{
			expr: "last day",
			want: TimePeriod{
				Period:   "day",
				Offset:   -1,
				Value:    1,
				Complete: false,
			},
		},
		{
			expr: "last complete day",
			want: TimePeriod{
				Period:   "day",
				Offset:   -1,
				Value:    1,
				Complete: true,
			},
		},
		{
			expr: "last 2 years",
			want: TimePeriod{
				Period:   "year",
				Offset:   -2,
				Value:    2,
				Complete: false,
			},
		},
		{
			expr: "last 2 complete months",
			want: TimePeriod{
				Period:   "month",
				Offset:   -2,
				Value:    2,
				Complete: true,
			},
		},
		{
			expr: "next 5 complete months",
			want: TimePeriod{
				Period:   "month",
				Offset:   1,
				Value:    5,
				Complete: true,
			},
		},
		{
			expr: "next 5 MONTHS",
			want: TimePeriod{
				Period:   "month",
				Offset:   0,
				Value:    5,
				Complete: false,
			},
		},
		{
			expr: "now",
			want: TimePeriod{
				Period:   "",
				Offset:   0,
				Value:    0,
				Complete: false,
			},
		},
		{
			expr: "today",
			want: TimePeriod{
				Period:   "day",
				Offset:   0,
				Value:    1,
				Complete: true,
			},
		},
		{
			expr: "tomorrow",
			want: TimePeriod{
				Period:   "day",
				Offset:   1,
				Value:    1,
				Complete: true,
			},
		},
		{
			expr: "yesterday     ",
			want: TimePeriod{
				Period:   "day",
				Offset:   -1,
				Value:    1,
				Complete: true,
			},
		},
		{
			expr:    "5 days",
			wantErr: true,
		},
		{
			expr:    "this  complete day",
			wantErr: true,
		},
		{
			expr:    "next something wrong",
			wantErr: true,
		},
		{
			expr:    "next 0 days",
			wantErr: true,
		},
		{
			expr:    "next -10 months",
			wantErr: true,
		},
		{
			expr:    "today 3 months",
			wantErr: true,
		},
	}
	for _, tt := range tests {
		t.Run(tt.expr, func(t *testing.T) {
			got, err := Parse(tt.expr)
			if (err != nil) != tt.wantErr {
				t.Errorf("parseString() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if !reflect.DeepEqual(got, tt.want) {
				t.Errorf("parseString() = %v, want %v", got, tt.want)
			}
		})
	}
}

func TestTimePeriod_IsTimezoneRelative(t *testing.T) {
	tests := []struct {
		expr string
		want bool
	}{
		{
			expr: "this month",
			want: true,
		},
		{
			expr: "this year",
			want: true,
		},
		{
			expr: "this hour",
			want: false,
		},
		{
			expr: "yesterday",
			want: true,
		},
		{
			expr: "last month",
			want: false,
		},
		{
			expr: "next complete week",
			want: true,
		},
		{
			expr: "next 5 complete hours",
			want: false,
		},
		{
			expr: "next 5 complete days",
			want: true,
		},
	}
	for _, tt := range tests {
		t.Run(tt.expr, func(t *testing.T) {
			tp, err := Parse(tt.expr)
			if err != nil {
				t.Error("failed parsing expression: %w", err)
			}
			if got := tp.IsTimezoneRelative(); got != tt.want {
				t.Errorf("TimePeriod.IsTimezoneRelative() %s => %v, want %v", tt.expr, got, tt.want)
			}
		})
	}
}

File: packages/functions-runtime/src/TimePeriod.test.js
import { test, expect } from "vitest";
const { TimePeriod } = require("./TimePeriod");

test("shorthands test", async () => {
  const today = TimePeriod.fromExpression("today");
  expect(today).toEqual({
    period: "day",
    value: 1,
    complete: true,
    offset: 0,
  });

  const tomorrow = TimePeriod.fromExpression("tomorrow");
  expect(tomorrow).toEqual({
    period: "day",
    value: 1,
    complete: true,
    offset: 1,
  });

  const yesterday = TimePeriod.fromExpression("yesterday");
  expect(yesterday).toEqual({
    period: "day",
    value: 1,
    complete: true,
    offset: -1,
  });

  const now = TimePeriod.fromExpression("now");
  expect(now).toEqual({
    period: "",
    value: 0,
    complete: false,
    offset: 0,
  });
  expect(now.periodStartSQL()).toEqual("(NOW())");
  expect(now.periodEndSQL()).toEqual("(NOW())");
});

test("next test", async () => {
  let period = TimePeriod.fromExpression("next day");
  expect(period).toEqual({
    period: "day",
    value: 1,
    complete: false,
    offset: 0,
  });

  period = TimePeriod.fromExpression("next complete day");
  expect(period).toEqual({
    period: "day",
    value: 1,
    complete: true,
    offset: 1,
  });

  period = TimePeriod.fromExpression("next 5 complete day");
  expect(period).toEqual({
    period: "day",
    value: 5,
    complete: true,
    offset: 1,
  });

  period = TimePeriod.fromExpression("next 5 months");
  expect(period).toEqual({
    period: "month",
    value: 5,
    complete: false,
    offset: 0,
  });

  period = TimePeriod.fromExpression("next 2 complete years");
  expect(period).toEqual({
    period: "year",
    value: 2,
    complete: true,
    offset: 1,
  });
});

test("last test", async () => {
  let period = TimePeriod.fromExpression("last day");
  expect(period).toEqual({
    period: "day",
    value: 1,
    complete: false,
    offset: -1,
  });

  period = TimePeriod.fromExpression("last complete day");
  expect(period).toEqual({
    period: "day",
    value: 1,
    complete: true,
    offset: -1,
  });

  period = TimePeriod.fromExpression("last 5 complete day");
  expect(period).toEqual({
    period: "day",
    value: 5,
    complete: true,
    offset: -5,
  });

  period = TimePeriod.fromExpression("last 5 months");
  expect(period).toEqual({
    period: "month",
    value: 5,
    complete: false,
    offset: -5,
  });

  period = TimePeriod.fromExpression("last 2 complete years");
  expect(period).toEqual({
    period: "year",
    value: 2,
    complete: true,
    offset: -2,
  });

  period = TimePeriod.fromExpression("last complete year");
  expect(period).toEqual({
    period: "year",
    value: 1,
    complete: true,
    offset: -1,
  });
});

test("errors test", async () => {
  expect(() => TimePeriod.fromExpression("last test day")).toThrowError(
    "Invalid time period expression"
  );
  expect(() => TimePeriod.fromExpression("today now")).toThrowError(
    "Invalid time period expression"
  );
  expect(() => TimePeriod.fromExpression("last -5 days")).toThrowError(
    "Invalid time period expression"
  );
  expect(() => TimePeriod.fromExpression("5 mont")).toThrowError(
    "Invalid time period expression"
  );
  expect(() => TimePeriod.fromExpression("5 days")).toThrowError(
    "Time period expression must start with this, next, or last"
  );
});
